h/api/models.py

```diff
@@ -166,6 +176,76 @@ def title(self):
             return ""
 
     @property
+    def filename(self):
+        if self.uri and self.uri.startswith("file://"):
+            return self.uri.split("/")[-1] or None
+
+    @property
+    def document_link(self):
+        """Return a link to this annotation's document.
```

Comment:
Ok, having looked into this, I want to keep it how it is. I understand that rendering HTML tags in Python instead of in the template is weird, but:

- This Python function is actually pretty complicated logic, and it's nice to keep it all in one place where it can be tested
- This logic is duplicated in JavaScript code for the annotation cards and now in Python for the page. With the code in Python, we can potentially add the link to the annotation on its way out of our API (in our search lib's `render()` function) and have the client-side code just use it. The problem with this though is annotations that are being created but haven't been saved to the server yet.

If we wanted to move this into the template, I think we'd need four separate Python functions to get the href, title, link_text, and domain, and then template code that looked like this:

    <a {% if href %}href="{{ href }}"{% endif %} title={{ title }}>{{ link_text }}</a>{% if domain %}({{ domain }}){% endif %}

And then even if we added the href, title, link_text, and domain to the annotations in `render()` we'd still have to duplicate that bit of template code containing (admittedly simple) logic client-side.

And splitting this one function into four would not be that simple as the values are inter-dependent (although having four smaller functions would probably benefit the code).

I'm willing to be overridden here though